contracts/zero-ex/contracts/src/features/TransformERC20.sol

```diff
@@ -0,0 +1,365 @@
+        }
+
+        IPuppet puppet = getTransformPuppet();
+
+        // Remember the initial output token balance of the taker.
+        uint256 takerOutputTokenBalanceBefore = minOutputTokenAmount == 0
+            ? 0 : LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+
+        // Pull input tokens from the taker to the puppet and transfer attached ETH.
+        _transferInputTokensAndAttachedEth(inputToken, taker, address(puppet), inputTokenAmount);
+
+        // Perform transformations.
+        for (uint256 i = 0; i < transformations.length; ++i) {
+            _executeTransformation(puppet, transformations[i], taker, callDataHash);
+        }
+
+        // Ensure enough output token has been sent to the taker.
+        if (minOutputTokenAmount != 0) {
+            // Compute how much output token has been transferred to the taker.
+            uint256 takerOutputTokenBalanceAfter =
+                LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+            if (takerOutputTokenBalanceAfter > takerOutputTokenBalanceBefore) {
+                outputTokenAmount = takerOutputTokenBalanceAfter.safeSub(
+                    takerOutputTokenBalanceBefore
+                );
+            }
+            if (outputTokenAmount < minOutputTokenAmount) {
+                LibTransformERC20RichErrors.IncompleteERC20TransformError(
+                    address(outputToken),
+                    outputTokenAmount,
+                    minOutputTokenAmount
+                ).rrevert();
+            }
+        }
+
+        // Emit an event.
+        emit TransformedERC20(
+            taker,
+            address(inputToken),
+            address(outputToken),
+            inputTokenAmount,
+            outputTokenAmount
+        );
+    }
+
+    /// @dev Return the current puppet instance that will serve as the execution
+    ///      context for transformations.
+    /// @return puppet The puppet instance.
+    function getTransformPuppet()
+        public
+        override
+        view
+        returns (IPuppet puppet)
+    {
+        return LibTransformERC20Storage.getStorage().puppet;
+    }
+
+    /// @dev Transfer input tokens from the taker and any attached ETH to `to`
+    /// @param inputToken The token to pull from the taker.
+    /// @param taker The taker address.
+    /// @param to The recipient of taker tokens and ETH.
+    /// @param amount Amount of `inputToken` tokens to transfer.
+    function _transferInputTokensAndAttachedEth(
+        IERC20TokenV06 inputToken,
+        address taker,
+        address payable to,
+        uint256 amount
+    )
+        private
+    {
+        // Transfer any attached ETH.
+        if (msg.value != 0) {
+            to.transfer(msg.value);
+        }
+        // Transfer input tokens.
+        if (!LibERC20Transformer.isTokenETH(inputToken)) {
+            // Token is not ETH, so pull the ERC20 from the taker.
+            ITokenSpender(address(this))._spendERC20Tokens(
```

Comment:
I might consider using the conventional `transferFrom` name here.

Reply:
That's kind of assuming that TokenSpender will only ever support ERC20s, right? I was thinking it would be upgraded to support other token types so I chose a descriptive name.